onnx/defs/shape_inference.h

```diff
@@ -172,94 +176,108 @@ inline void updateOutputElemType(
       (output_type->value_case() == TypeProto::kTensorType ||
        output_type->value_case() == TypeProto::VALUE_NOT_SET)) {
     output_type->mutable_tensor_type()->set_elem_type(elemType);
+	return Status::OK();
   }
+  return Status(INFERENCE, INVALID_PROTOBUF, "Update output elem type failed. output_type should be a tensor-type.");
 }
 
 // Infer type of an output from the value of a specified attribute, which is expected
 // to have a valid value representing a TensorProto_DataType.
-inline void propagateElemTypeFromAttributeToOutput(
+inline Status propagateElemTypeFromAttributeToOutput(
     InferenceContext& ctx,
     const std::string& attributeName,
     size_t outputIndex,
     TensorProto_DataType default_value = TensorProto::UNDEFINED) {
   auto attr_proto = ctx.getAttribute(attributeName);
   if (nullptr == attr_proto) { // attribute not present
-    if (default_value != TensorProto::UNDEFINED)
-      updateOutputElemType(ctx, outputIndex, default_value);
-    return;
+    if (default_value != TensorProto::UNDEFINED) {
+      return updateOutputElemType(ctx, outputIndex, default_value);
+	}
+    return Status(INFERENCE, INVALID_PROTOBUF, MakeString("Attribute: ", attributeName, " does not exist."));
+  }
+  if (!attr_proto->has_i()) {
+	  // attribute not of right type
+	  return Status(INFERENCE, INVALID_PROTOBUF, MakeString("Attribute: ", attributeName, " should be integer type."));
   }
-  if (!attr_proto->has_i()) return; // attribute not of right type
   auto attr_value = attr_proto->i();
   auto elem_type = static_cast<TensorProto_DataType>(attr_value);
-  if (!TensorProto_DataType_IsValid(elem_type)) return; // out-of-range attribute value
-  updateOutputElemType(ctx, outputIndex, elem_type);
+  if (!TensorProto_DataType_IsValid(elem_type)) {
+	  return Status(INFERENCE, INVALID_PROTOBUF, MakeString("Attribute: ", attributeName, " is not specifying a valid tensor element type."));
+  }
+  return updateOutputElemType(ctx, outputIndex, elem_type);
 }
 
 inline TensorShapeProto* getOutputShape(InferenceContext& ctx, size_t n) {
   auto output_type = ctx.getOutputType(n);
   if ((output_type != nullptr) &&
       (output_type->value_case() == TypeProto::kTensorType ||
        output_type->value_case() == TypeProto::VALUE_NOT_SET)) {
-         return output_type->mutable_tensor_type()->mutable_shape();
-	} else
-      return nullptr;
+      return output_type->mutable_tensor_type()->mutable_shape();
+  }
+  else {
+	  return nullptr;
+  }
 }
 
-inline void updateOutputShape(
+inline Status updateOutputShape(
     InferenceContext& ctx,
     size_t outputIndex,
     const TensorShapeProto& shape) {
   auto* output_shape = getOutputShape(ctx, outputIndex);
-  if (output_shape != nullptr) {
-    *output_shape = shape;
+  if (output_shape == nullptr) {
+	return Status(INFERENCE, INVALID_PROTOBUF, MakeString("Output: ", outputIndex, " should be tensor-type or no-set."));
```

Comment:
I think this won't work. We need to allow optional outputs (which may be missing, and that is valid). Unfortunately, we can distinguish between the two cases only inside "getOutputShape".

Reply:
Ok, please ignore my comment above, as per discussion elsewhere

Reply:
But, essentially, this condition will never arise as per our discussion.